Add SelectiveStreamReaders for BIGINT, INTEGER, SMALLINT and DATE types

mbasmanova opened the pull request:
Depends on #12991

Part of #12585

mbasmanova commented:
Incorporated feedback from #12991 and rebased onto master.

mbasmanova commented:
@highker 

> "Fix memory accounding in LongDictionaryStreamReader" except two types in the commit title

James, thank you for reviewing. Typos fixed.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDirectSelectiveStreamReader.java:
```diff
@@ -0,0 +1,245 @@
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
```
Even `outputRequired` implies `outputType` is not null, maybe it's still good to have `checkState(outputType != null, ...)`

mbasmanova replied:
I don't think an extra check is necessary. These variables are managed locally and therefore it is safe to assume what their values are. If outputRequired is true, but outputType is null, then there is a bug in this class.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDirectSelectiveStreamReader.java:
```diff
@@ -0,0 +1,245 @@
+
+    private boolean allNulls;
+
+    private LocalMemoryContext systemMemoryContext;
```
Seems this doesn't do anything? If this is redundant, may could simplify `close()` logic.

mbasmanova replied:
@highker James, thanks for pointing this out. `systemMemoryContext` is not needed for BatchStreamReaders because they don't keep any *state* - I'll make a PR to remove `systemMemoryContext` from all XxxBatchStreamReaders. However, SelectiveStreamReaders re-use arrays and therefore must account for their memory properly. I'll add the necessary logic. 

CC: @yingsu00

mbasmanova replied:
Added logic to update memory usage in `read` methods.

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongSelectiveStreamReader.java:
```diff
@@ -0,0 +1,130 @@
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            Optional<Type> outputType,
+            AggregatedMemoryContext systemMemoryContext)
```
Same for this one and the one for dictionary reader; not used?

highker commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongSelectiveStreamReader.java:
```diff
@@ -0,0 +1,209 @@
+        return outputPositions;
+    }
+
+    protected Block makeOutputBlock(int[] positions, int positionCount, boolean includeNulls)
```
`buildOutputBlock` maybe

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -61,4 +80,171 @@ public void testBooleanSequence()
+        return ContiguousSet.create(Range.closedOpen(lowerInclusive, upperExclusive), DiscreteDomain.integers());
+    }
+
+    private static <T> Iterable<T> skipEvery(int n, Iterable<T> iterable)
```
I think `n` is appropriate here as this method skips every n-th row.

mbasmanova commented:
@highker James, thank you for review. I added logic to account for memory used by `values`, `nulls`, and `outputPositions` arrays; renamed `makeOutputBlock` to `buildOutputBlock`; changed `close` method to use lambdas. PR is updated.

tdcmeehan commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:
```diff
@@ -179,17 +175,24 @@ public OrcSelectiveRecordReader createRecordReader(Optional<TupleDomainFilter> f
-        @Param("boolean")
+        @Param({
+                "boolean",
+
```
Nit: extra space

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java:
```diff
@@ -128,7 +128,7 @@ public OrcSelectiveRecordReader(
                         streamDescriptor,
                         Optional.ofNullable(filters.get(columnId)),
-                        outputColumns.contains(columnId),
+                        outputColumns.contains(columnId) ? Optional.of(includedColumns.get(columnId)) : Optional.empty(),
```
Would do `Optional.ofNullable(includedColumns.get(columnId)),` consistent with line above

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractLongSelectiveStreamReader.java:
```diff
@@ -0,0 +1,209 @@
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+abstract class AbstractLongSelectiveStreamReader
```
This is just a minor naming problem, but it's a little confusing to me that `LongSelectiveStreamReader` doesn't extend from `AbstractLongSelectiveStreamReader`.

mbasmanova replied:
@tdcmeehan That's an interesting observation. I cannot think of a solution though.

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongSelectiveStreamReader.java:
```diff
@@ -0,0 +1,130 @@
+        ColumnEncoding.ColumnEncodingKind kind = encoding.get(streamDescriptor.getStreamId())
+                .getColumnEncoding(streamDescriptor.getSequence())
+                .getColumnEncodingKind();
+        if (kind == DIRECT || kind == DIRECT_V2 || kind == DWRF_DIRECT) {
```
I think this would be better expressed as a switch statement

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/LongSelectiveStreamReader.java:
```diff
@@ -0,0 +1,130 @@
+            Optional<Type> outputType,
+            AggregatedMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
```
`requireNonNull(streamDescriptor, "streamDescriptor is null");`

mbasmanova commented:
@tdcmeehan Tim, thank you for review. Comments addressed. Rebased onto latest master.
